docs/proposals/configmap-rollout.md

```diff
@@ -0,0 +1,171 @@
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/kubernetes/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# ConfigMap management
+
+## Abstract
+
+A proposal for implementing management of ConfigMaps in the context of a
+[Deployment](http://kubernetes.io/docs/user-guide/deployments/).
+
+Users should be able to have their apps automatically pick up changes to
+their configuration and also have the ability to rollback to previous
+configuration similar to the ability to rollback to previous pod template
+revisions.
+
+Currently, in order to apply updated configuration into a running app, users
+need to 1) either do it manually: copy the ConfigMap that holds their old
+behavior, update it, store it back as a different ConfigMap, and mount the
+new ConfigMap into the Deployment, or 2) edit the original ConfigMap and
+trigger the deployment controller to start a new ReplicaSet (either by
+adding a dummy environment value in the pod template of the Deployment or
+by including the update as part of another update) losing the ability to
+rollback to previous configuration.
+
+This doc proposes a solution for removing the burden of ConfigMap management
+from users by providing a mechanism that will work out of the box without
+having to manually rollout new ReplicaSets.
+
+## Implementation
+
+### Proposed solution
+
+Users should only ever need to create and update a single ConfigMap "foo".
+
+```sh
+$ kubectl create configmap foo --from-file=config.yml
+configmap "foo" created
+```
+
+We need a way to link "foo" to an owner. Owner references in ConfigMaps
+should enable their management. ConfigMaps without owner references will
+continue to work as expected today.
+
+```sh
+$ kubectl set owner configmap/foo deployment/bar
+configmap "foo" updated
+```
+
+The ConfigMap needs to be mounted as a volume in the Deployment it references as an owner.
+
+```sh
+$ kubectl set volume deployment/bar -t configmap --configmapname foo
+deployment "bar" updated
+```
+
+A new controller needs to watch for Deployments and ConfigMaps. Whenever it
+spots a Deployment with a mounted ConfigMap that has an owner reference
+pointing back to the Deployment, it will create a copy of the ConfigMap, 
+remove the original owner reference, and mount the copy in the owner. It 
```

Comment:
> No, because "Foo" always has the owner reference to the Deployment.

The relationship has to be two-way... the configmap points to the deployment, and the deployment references the configmap in a particular volume definition (or envvar valueFrom spec). If the deployment is rewritten to reference the latest configmap, the connection from the deployment to the original configmap is lost